File: FillBlankQuestionDemo/app/src/main/java/com/yl/fillblankquestiondemo/MainActivity.java
package com.yl.fillblankquestiondemo;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * 填空题
 * Created by yangle on 2017/9/2.
 * <p>
 * Website：http://www.yangle.tech
 * <p>
 * GitHub：https://github.com/alidili
 * <p>
 * CSDN：http://blog.csdn.net/kong_gu_you_lan
 * <p>
 * JianShu：http://www.jianshu.com/u/34ece31cd6eb
 */

public class MainActivity extends AppCompatActivity {

    @BindView(R.id.fbv_content)
    FillBlankView fbvContent;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);

        initData();
    }

    private void initData() {
        String content = "纷纷扬扬的________下了半尺多厚。天地间________的一片。我顺着________工地走了四十多公里，" +
                "只听见各种机器的吼声，可是看不见人影，也看不见工点。一进灵官峡，我就心里发慌。";

        // 答案范围集合
        List<AnswerRange> rangeList = new ArrayList<>();
        rangeList.add(new AnswerRange(5, 13));
        rangeList.add(new AnswerRange(23, 31));
        rangeList.add(new AnswerRange(38, 46));

        fbvContent.setData(content, rangeList);
    }
}

File: AshmemDemo/app/src/main/java/com/yangle/ashmem/MainActivity.kt
package com.yangle.ashmem

import android.content.ComponentName
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.IBinder
import android.os.ParcelFileDescriptor
import androidx.appcompat.app.AppCompatActivity
import java.io.File
import java.io.FileOutputStream

/**
 * 共享内存传输数据
 * Created by yangle on 2025/12/15.
 * <p>
 * Website：http://www.yangle.tech
 * GitHub：https://github.com/alidili
 * CSDN：http://blog.csdn.net/kong_gu_you_lan
 * JianShu：http://www.jianshu.com/u/34ece31cd6eb
 */
class MainActivity : AppCompatActivity() {

    private val native = NativeShm()
    private lateinit var service: IShmService

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        bindService(Intent(this, ShmService::class.java), conn, BIND_AUTO_CREATE)
    }

    private val conn = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName, binder: IBinder) {
            service = IShmService.Stub.asInterface(binder)
            // 通知服务端开始发送文件
            service.startTransfer(object : IShmCallback.Stub() {
                override fun onShmReady(pfd: ParcelFileDescriptor) {
                    startReceive(pfd.fd)
                }
            })
        }

        override fun onServiceDisconnected(name: ComponentName) {}
    }

    /**
     * 接收文件
     *
     * @param fd 文件描述符
     */
    private fun startReceive(fd: Int) {
        Thread {
            val out = File(getExternalFilesDir(""), "test.jpg")
            if (out.exists()) {
                out.delete()
            }
            val buf = ByteArray(64 * 1024)

            FileOutputStream(out).use { fos ->
                while (true) {
                    val len = native.read(fd, buf)
                    if (len < 0) {
                        service.endTransfer()
                        break
                    }
                    fos.write(buf, 0, len)
                }
            }
        }.start()
    }
}